### PR TITLE
Update cse_opts.py

### LINKED ISSUE #19136: New Hang in simplify()
The following code runs in under ~4 seconds using SymPy 1.5.1 or earlier (at least since SymPy 0.7.6), but when run using latest SymPy-dev (as of April 15, 2020) hangs indefinitely:

https://gist.github.com/zachetienne/bda46f9d9ca3b96ba3d612a9b8b7d69a

I caught this hang because Travis timed out on routine CI unit tests for my code NRPy+ when using the latest SymPy-dev (https://github.com/zachetienne/nrpytutorial ; http://nrpyplus.net/). The Travis build at 10:24AM Eastern time on April 15 (https://travis-ci.org/github/zachetienne/nrpytutorial/builds/675286260) completed successfully, but later builds failed. This should help pinpoint the precise SymPy commit that caused this bug.

### PR DESCRIPTION
**Pull request by iammosespaulr**
#### References to other Issues or PRs

fix attempt for #19136

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by Sc0rpi0n101** [comment]
Since the bug was introduced only a few days ago, and it will be fixed without going through a release, I don't think there's a need for release notes.

**Comment by oscarbenjamin** [comment]
I don't think that this should say that it "fixes #19136" because the cause of the issue is still not identified and needs to be investigated.

**Comment by iammosespaulr** [comment]
> I don't think that this should say that it "fixes #19136" because the cause of the issue is still not identified and needs to be investigated.

I converted it to a draft, till we figure that out

**Comment by smichr** [comment]
> the cause of the issue is still not identified

Add/Mul, even when unevaluated, will remove the cls.identity:

```
>>> Mul(1,-1,2,evaluate=False).args
(-1, 2)
>>> Mul._from_args(Tuple(1,-1,2)).args
(1, -1, 2)
>>> Add(0,-1,2,evaluate=False).args
(-1, 2)
>>> Add._from_args(Tuple(0,-1,2)).args
(0, -1, 2)
```

So to create this special form in cse, the `_from_args` must be used. Why that only now is exhibiting the issue, I do not know.

**Comment by iammosespaulr** [comment]
> > the cause of the issue is still not identified
> 
> Add/Mul, even when unevaluated, will remove the cls.identity:
> 
> ```
> >>> Mul(1,-1,2,evaluate=False).args
> (-1, 2)
> >>> Mul._from_args(Tuple(1,-1,2)).args
> (1, -1, 2)
> >>> Add(0,-1,2,evaluate=False).args
> (-1, 2)
> >>> Add._from_args(Tuple(0,-1,2)).args
> (0, -1, 2)
> ```
> 
> So to create this special form in cse, the `_from_args` must be used. Why that only now is exhibiting the issue, I do not know.

Ohhh I see, so I tried executing that script with 
```python3
elif a.could_extract_minus_sign():
                negs[a] = Mul._from_args([S.NegativeOne, -a])
```
and it hung.
and then I checked if they were equivalent
```python3
>>> Mul(S.One, S.NegativeOne, -a, evaluate=False) == Mul._from_args([S.NegativeOne, -a])
True
>>> 
``` 
They are, so maybe `Mul(blah)` isn't really suitable for simplifying cause it leaves the identity behind ? 🤷

**Comment by oscarbenjamin** [comment]
> Why that only now is exhibiting the issue, I do not know.

It is possible that a slight change in the form of the expression has exposed a bug somewhere else that leads to simplify hanging. If so then the code that hangs is the real problem that needs to be fixed.

**Comment by smichr** [comment]
> then the code that hangs is the real problem

maybe -- but the circuitry of cse being able to efficiently mark the expressions of interest is very important. When I studied this years ago I was amazed at how fast it was and how flexible it was. I am not sure that the unevaluated Mul is as good and that in itself is perhaps a regression.

**Comment by iammosespaulr** [comment]
Here're some differences between the the code that hangs and the one that doesn't
This might help ?

I had `negs` printed out 
```python3
#Print negs hung

{g4DD11*g4DD22*g4DD33 - g4DD11*g4DD23**2 - g4DD12**2*g4DD33 + ... } x 1
{} x 24
- Hung

#Print negs not hanging

{g4DD11*g4DD22*g4DD33 - g4DD11*g4DD23**2 - g4DD12**2*g4DD33 + ... } x 1
{AMPLRHO**4*AMPLZ**2*g4DD11*g4DD22*g4DD33*(exp(xx0/SINHWRHO)/SINHWRHO + ... } x 6
{} x 27
3
```

**Comment by oscarbenjamin** [comment]
> I am not sure that the unevaluated Mul is as good and that in itself is perhaps a regression.

If that's the case then maybe some of the other changes from e8fa81e3544a91af6bada966dd556c45bbcfc9a0 or #19083 should be reconsidered as well.

File: sympy/simplify/cse_opts.py [hunk 1]
```diff
@@ -14,7 +14,14 @@ def sub_pre(e):
     """
     # replacing Add, A, from which -1 can be extracted with -1*-A
     adds = [a for a in e.atoms(Add) if a.could_extract_minus_sign()]
-    reps = dict((a, Mul(S.NegativeOne, -a, evaluate=False)) for a in adds)
+    reps = {}
+    ignore = set()
+    for a in adds:
+        na = -a
+        if na.is_Mul:  # e.g. MatExpr
+            set.add(a)
```
Comment (smichr) [inline sympy/simplify/cse_opts.py:22]:
```suggestion
            ignore.add(a)
```

**Comment by smichr** [comment]
We should find a test that demonstrates the need for the class identity to remain as a marker for the Adds that needed to be handled in the 2nd pass.

**Comment by smichr** [comment]
> should be reconsidered as well.

We should add a test that should not regress in terms of the time taken for it, too. I don't recall such an expression now, but I do recall that the time difference was quite dramatic for some large expression.

**Comment by smichr** [comment]
> find a test that demonstrates the need for the class identity to remain as a marker

Well...the issue gives such an expression. It would be nice to have a smaller failing example, however.

**Comment by zachetienne** [comment]
While the true source of the strange behavior is found, this one-line patch will restore the previous (desirable) behavior for my use case. 

It seems my Travis runs may be useful for catching future SymPy-dev bugs, so I'd like to resume them so in case there's another issue in the meantime I might be able to catch it. But this patch would need to be applied for this to happen.

Given that this one-line-patch PR fixes the issue in question (#19136) without causing any breakages, would it be possible to get this PR reviewed & merged to master in the meantime?

**Comment by smichr** [comment]
> PR reviewed & merged to master in the meantime?

tests are in process

**Comment by smichr** [comment]
The previous test which motivated the changes that broke the sub_pre/post processing still passes while applying a change that retains the pre-change use of `_from_args`. Tests can be added in a new PR; the issue will be updated.

**Comment by smichr** [comment]
> I don't think that this should say that it "fixes #19136"

By saying that, it caused the issue to close when this was merged even though that was not in the OP anymore. This should be reported to github.

**Comment by oscarbenjamin** [comment]
I've never been clear what does the "fixes" matching. Is it a standard Github feature or is it part of a bot that has been specifically enabled on sympy/sympy?

**Comment by iammosespaulr** [comment]
> I've never been clear what does the "fixes" matching. Is it a standard Github feature or is it part of a bot that has been specifically enabled on sympy/sympy?

GitHub, 
[When you merge a linked pull request into the default branch of a repository, its linked issue is automatically closed.](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue)

**Pull request merged**